### PR TITLE
ci: Optimize workflows and use ruff for linting

### PR DESCRIPTION
**Pull request by jvirgovic**
## Description

- Updated some GitHub actions to be up 2 date.
- Removed flake8 linter from Continuous integration and created standalone yaml file for this + Replaced flake8 for [Ruff](https://github.com/astral-sh/ruff), which is much faster.
- Some other nitpicks..

## Motivation and Context

n/a

## How Has This Been Tested?

n/a

## Screenshots (if appropriate):
n/a

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by jvirgovic** [comment]
@orhun This should be ready to merge right away.

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    runs-on: ubuntu-22.04
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Ruff
+        uses: chartboost/ruff-action@v1
+        with:
+          args: check .
+          config: ruff.toml
```
Comment (orhun) [inline .github/workflows/lint.yml]: There is a warning in the checks:

> Unexpected input(s) 'config', valid inputs are ['args', 'src', 'version']

Are we sure that the action supports `config`?

File: ruff.toml [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+    "env",
+    "__pycache__",
+    ".git",
+]
+
+indent-width = 4
+target-version = "py310"
+
+[lint]
+ignore = ["E501", "E226"]
```
Comment (orhun) [inline ruff.toml]: Can you add comments explaining which lints are ignored?

**Review by orhun — COMMENTED** [review]
Nice, I like it. Left some comments.

File: .github/workflows/lint.yml (hunk 1 above)
Comment (jvirgovic) [reply .github/workflows/lint.yml]: My bad, it's not needed since Ruff by default looks for ruff.toml.

File: ruff.toml (hunk 2 above)
Comment (jvirgovic) [reply ruff.toml]: Done.

**Review by orhun — APPROVED** [review]
Thanks!

**Pull request merged**